### PR TITLE
[Feature][Transform-V2] Support transform with multi-table

### PR DESCRIPTION
**Pull request by Hisoka-X**
### Purpose of this pull request
Support transform with multi-table

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
add new test.

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  2. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
  3. Add ci label in [label-scope-conf](https://github.com/apache/seatunnel/blob/dev/.github/workflows/labeler/label-scope-conf.yml)
  4. Add e2e testcase in [seatunnel-e2e](https://github.com/apache/seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/)
  5. Update connector [plugin_config](https://github.com/apache/seatunnel/blob/dev/config/plugin_config)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by Hisoka-X** [comment]
waiting https://github.com/apache/seatunnel/pull/7687

**Comment by cobolbaby** [comment]
Can we move forward now?

**Comment by Hisoka-X** [comment]
cc @hailin0 and @liunaijie

File: seatunnel-core/seatunnel-flink-starter/seatunnel-flink-starter-common/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformExecuteProcessor.java [hunk 1]
```diff
@@ -130,7 +129,7 @@ public List<DataStreamTableInfo> execute(List<DataStreamTableInfo> upstreamDataS
                         resultTableName,
                         new DataStreamTableInfo(
                                 inputStream,
-                                Collections.singletonList(transform.getProducedCatalogTable()),
+                                transform.getProducedCatalogTables(),
```
Comment (hailin0) [inline seatunnel-core/seatunnel-flink-starter/seatunnel-flink-starter-common/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformExecuteProcessor.java:130]: Do spark/zeta need to be modified?

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/flow/TransformFlowLifeCycle.java [hunk 2]
```diff
@@ -89,6 +90,20 @@ public void received(Record<?> record) {
...
             collector.collect(record);
+        } else if (record.getData() instanceof SchemaChangeEvent) {
+            if (prepareClose) {
+                return;
+            }
+            SchemaChangeEvent event = (SchemaChangeEvent) record.getData();
+            for (SeaTunnelTransform<T> t : transform) {
+                event = t.mapSchemaChangeEvent(event);
+                if (event == null) {
+                    break;
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/flow/TransformFlowLifeCycle.java:106]: add log break details. e.g transform name、config...

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/flow/TransformFlowLifeCycle.java [hunk 3]
```diff
@@ -89,6 +90,20 @@ public void received(Record<?> record) {
             // ack after #addState
             runningTask.ack(barrier);
             collector.collect(record);
+        } else if (record.getData() instanceof SchemaChangeEvent) {
+            if (prepareClose) {
+                return;
+            }
+            SchemaChangeEvent event = (SchemaChangeEvent) record.getData();
+            for (SeaTunnelTransform<T> t : transform) {
+                event = t.mapSchemaChangeEvent(event);
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/flow/TransformFlowLifeCycle.java]: log print event before/after

File: seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-1/src/test/java/org/apache/seatunnel/e2e/transform/TestCopyIT.java [hunk 4]
```diff
@@ -32,4 +34,15 @@ public void testCopy(TestContainer container) throws IOException, InterruptedExc
         Container.ExecResult execResult = container.executeJob("/copy_transform.conf");
         Assertions.assertEquals(0, execResult.getExitCode());
     }
+
+    @DisabledOnContainer(
+            value = {},
+            type = {EngineType.SPARK},
+            disabledReason = "Currently SPARK do not multi table transform")
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-1/src/test/java/org/apache/seatunnel/e2e/transform/TestCopyIT.java:41]: Why?

Comment (hailin0) [reply seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-1/src/test/java/org/apache/seatunnel/e2e/transform/TestCopyIT.java:41]: Or please create an issue

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/common/AbstractMultiCatalogFlatMapTransform.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+package org.apache.seatunnel.transform.common;
+
+import org.apache.seatunnel.api.configuration.ReadonlyConfig;
+import org.apache.seatunnel.api.table.catalog.CatalogTable;
+import org.apache.seatunnel.api.table.type.SeaTunnelRow;
+import org.apache.seatunnel.api.transform.SeaTunnelFlatMapTransform;
+
+import java.util.List;
+
+public abstract class AbstractMultiCatalogFlatMapTransform extends AbstractMultiCatalogTransform
```
Comment (hailin0) [inline seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/common/AbstractMultiCatalogFlatMapTransform.java:28]: Add code comments

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/common/AbstractMultiCatalogMapTransform.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+package org.apache.seatunnel.transform.common;
+
+import org.apache.seatunnel.api.configuration.ReadonlyConfig;
+import org.apache.seatunnel.api.table.catalog.CatalogTable;
+import org.apache.seatunnel.api.table.type.SeaTunnelRow;
+import org.apache.seatunnel.api.transform.SeaTunnelMapTransform;
+
+import java.util.List;
+
+public abstract class AbstractMultiCatalogMapTransform extends AbstractMultiCatalogTransform
```
Comment (hailin0) [inline seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/common/AbstractMultiCatalogMapTransform.java:28]: Add code comments

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/common/AbstractMultiCatalogTransform.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+import org.apache.seatunnel.api.transform.SeaTunnelTransform;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public abstract class AbstractMultiCatalogTransform implements SeaTunnelTransform<SeaTunnelRow> {
```
Comment (hailin0) [inline seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/common/AbstractMultiCatalogTransform.java:39]: Add code comments

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/common/TransformCommonOptions.java [hunk 8]
```diff
@@ -17,14 +17,37 @@
...
+public class TransformCommonOptions {
+
+    public static final Option<List<Map<String, Object>>> MULTI_TABLES =
+            Options.key("table_transform")
+                    .type(new TypeReference<List<Map<String, Object>>>() {})
+                    .defaultValue(Collections.emptyList())
+                    .withDescription("The table transform config");
 
-public final class CommonOptions {
+    public static final Option<String> TABLE_PATH =
```
Comment (hailin0) [inline seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/common/TransformCommonOptions.java:38]: add into docs

**Review by hailin0 — COMMENTED** [review]
update docs

https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/contribute-transform-v2-guide.md
https://github.com/apache/seatunnel/blob/dev/docs/zh/contribution/contribute-transform-v2-guide.md

File: seatunnel-core/seatunnel-flink-starter/seatunnel-flink-starter-common/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformExecuteProcessor.java (hunk 1 above)
Comment (Hisoka-X) [reply seatunnel-core/seatunnel-flink-starter/seatunnel-flink-starter-common/src/main/java/org/apache/seatunnel/core/starter/flink/execution/TransformExecuteProcessor.java:130]: spark not support at now. Zeta already support it in https://github.com/apache/seatunnel/pull/7628/files#diff-1c66b115c3fe7ea5fb8f48e9b51ed89fb9e58a2752846accc2df8ba663ce5861R505

File: seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-1/src/test/java/org/apache/seatunnel/e2e/transform/TestCopyIT.java (hunk 4 above)
Comment (Hisoka-X) [reply seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-1/src/test/java/org/apache/seatunnel/e2e/transform/TestCopyIT.java:41]: I will create an issue when this pr merged.

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/flow/TransformFlowLifeCycle.java (hunk 2 above)
Comment (Hisoka-X) [reply seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/flow/TransformFlowLifeCycle.java:106]: done

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/flow/TransformFlowLifeCycle.java (hunk 3 above)
Comment (Hisoka-X) [reply seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/flow/TransformFlowLifeCycle.java]: done

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/common/AbstractMultiCatalogFlatMapTransform.java (hunk 5 above)
Comment (Hisoka-X) [reply seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/common/AbstractMultiCatalogFlatMapTransform.java:28]: done

File: seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/common/TransformCommonOptions.java (hunk 8 above)
Comment (Hisoka-X) [reply seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/common/TransformCommonOptions.java:38]: done

File: seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-1/src/test/java/org/apache/seatunnel/e2e/transform/TestCopyIT.java (hunk 4 above)
Comment (Hisoka-X) [reply seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-1/src/test/java/org/apache/seatunnel/e2e/transform/TestCopyIT.java:41]: https://github.com/apache/seatunnel/issues/8138

**Review by hailin0 — APPROVED** [review]
LGTM

**Review by davidzollo — APPROVED** [review]
+1

**Pull request merged**